drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
+            full_result.success() & modules,
+            full_result.failure(),
+            full_result.error(),
+        )
+        .unwrap();
+
+        // Write one bitfield per active port in the ModuleId which was
+        // successfully retrieved above.
         let mut count = 0;
-        for mask in modules.ports.to_indices().map(|i| 1 << i) {
+        for mask in modules
+            .to_indices()
+            .map(|i| 1 << i)
+            .filter(|&m| desired_result.success().0 & m != 0)
+        {
             let mut status = Status::empty();
-            if (enable.get() & mask) != 0 {
+            if (mod_status.power_enable & mask) != 0 {
                 status |= Status::ENABLED;
             }
-            if (!resetl.get() & mask) != 0 {
+            if (!mod_status.resetl & mask) != 0 {
                 status |= Status::RESET;
             }
-            if (lpmode.get() & mask) != 0 {
+            if (mod_status.lpmode_txdis & mask) != 0 {
                 status |= Status::LOW_POWER_MODE;
             }
-            if (!presentl.get() & mask) != 0 {
+            if (!mod_status.modprsl & mask) != 0 {
                 status |= Status::PRESENT;
             }
-            if (!irql.get() & mask) != 0 {
+            if (!mod_status.intl_rxlosl & mask) != 0 {
                 status |= Status::INTERRUPT;
             }
-            if (pg.get() & mask) != 0 {
+            if (mod_status.power_good & mask) != 0 {
                 status |= Status::POWER_GOOD;
             }
-            if (pg_to.get() & mask) != 0 {
+            if (mod_status.power_good_timeout & mask) != 0 {
                 status |= Status::FAULT_POWER_TIMEOUT;
             }
-            if (pg_lost.get() & mask) != 0 {
+            if (mod_status.power_good_fault & mask) != 0 {
                 status |= Status::FAULT_POWER_LOST;
             }
             // Convert from Status -> u8 and write to the output buffer
             out[count] = status.bits();
             count += 1;
         }
-        Ok(count)
+        (count, desired_result)
     }
 
     fn select_page(
         &mut self,
         page: Page,
-        mask: FpgaPortMasks,
-    ) -> Result<(), HwError> {
+        mask: LogicalPortMask,
+    ) -> ModuleResult {
         // Common to both CMIS and SFF-8636
         const BANK_SELECT: u8 = 0x7E;
         const PAGE_SELECT: u8 = 0x7F;
 
+        let mut result = ModuleResult::default();
+
         // We can always write the lower page; upper pages require modifying
         // registers in the transceiver to select it.
         if let Some(page) = page.page() {
-            self.transceivers
-                .set_i2c_write_buffer(&[page])
-                .map_err(|_e| HwError::PageSelectWriteBufFailed)?;
-            self.transceivers
-                .setup_i2c_write(PAGE_SELECT, 1, mask)
-                .map_err(|_e| HwError::PageSelectWriteFailed)?;
-            self.wait_and_check_i2c(mask)?;
+            result =
```

Comment:
Same point here about chaining with `set_i2c_write_buffer()`. I think that will ensure that we almost always return all 1's as the successful modules from this method.